packages/store/addon/-private/system/model/shim-model-class.ts

```diff
@@ -0,0 +1,68 @@
+import { map } from '@ember/object/computed';
+import CoreStore from '../core-store';
+import { stringify } from 'querystring';
+import { RelationshipSchema, AttributeSchema } from '../../ts-interfaces/record-data-schemas';
+import { Dict } from '../../ts-interfaces/utils';
+
+type NormalDict<K extends string, V> = { [KK in K]: V | undefined };
+type ConfidentDict<K extends string, V> = { [KK in K]: V };
+
+type MakeConfident<D> = D extends NormalDict<infer K, infer T> ? ConfidentDict<K, T> : never;
+
+// Mimics the static apis of DSModel
+export default class ShimModelClass {
+  store: CoreStore;
+  modelName: string;
+  _id?: string;
+
+  // TODO Maybe expose the class here?
+  constructor(store: CoreStore, modelName: string, id?: string) {
+    this.store = store;
+    this.modelName = modelName;
+    this._id = id;
+  }
+
+  get fields(): Map<string, 'attribute' | 'belongsTo' | 'hasMany'> {
+    let attrs = this.store._attributesDefinitionFor(this.modelName, this._id);
+    let relationships = this.store._relationshipsDefinitionFor(this.modelName, this._id);
+    let fields = new Map<string, 'attribute' | 'belongsTo' | 'hasMany'>();
+    Object.keys(attrs).forEach(key => fields.set(key, 'attribute'));
+    Object.keys(relationships).forEach(key => fields.set(key, relationships[key]!.kind));
+    return fields;
+  }
+
+  get attributes(): Map<string, AttributeSchema> {
+    let attrs = this.store._attributesDefinitionFor(this.modelName, this._id);
+    return new Map(Object.entries(attrs) as [string, AttributeSchema][]);
+  }
+
+  get relationshipsByName(): Map<string, RelationshipSchema> {
+    let relationships = this.store._relationshipsDefinitionFor(this.modelName, this._id);
+    //TODO Talk to mnorth about needing a cast
+    return new Map(Object.entries(relationships as MakeConfident<RelationshipSchema>));
+    //return new Map(Object.entries(relationships) as [string, RelationshipSchema][]);
+  }
+
+  eachAttribute(callback: (key: string, attribute: AttributeSchema) => void, binding: unknown) {
```

Comment:
Might be overdoing it, but this is conceptually what the intent is 
```ts
eachAttribute<T>(callback: (this: T, key: string, attribute: AttributeSchema) => void, binding: T)
```

Reply:
this is pretty neat